kep/260-network-aware-scheduling/README.md

```diff
@@ -0,0 +1,1511 @@
+	// Standard list metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// Items is the list of AppGroup
+	Items []AppGroup `json:"items"`
+}
+```
+
+### AppGroup Test based on Online Boutique
+
+In this test, an AppGroup is created for the [Online Boutique application](https://github.com/GoogleCloudPlatform/microservices-demo).
+It consists of 11 workloads each one corresponding to a group of Pods, which we named from P1 - P11. 
+
+<p align="center"><img src="figs/appGroupTestOnlineBoutique.png" title="appGroupTestOnlineBoutique" width="800" class="center"/></p>
+
+As shown below, the preferred order for the KahnSort algorithm is P1, P10, P9, P8, P7, P6, P5, P4, P3, P2, P11. 
+
+We attribute an **index** to each workload to evaluate their topology preference in the **Less function of the TopologicalSort plugin** described [here](#description-of-the-topologicalsort).
+The topology list corresponds to:
+
+```go
+topologyList = [(P1 1) (P10 2) (P9 3) (P8 4) (P7 5) (P6 6) (P5 7) (P4 8) (P3 9) (P2 10) (P11 11)]
+```
+
+<p align="center"><img src="figs/appGroupTest.png" title="appGroupTest" width="900" class="center"/></p>
+
+## Network Topology CRD
+
+We also designed a NetworkTopology CRD based on the current [NodeResourceTopology CRD](https://github.com/kubernetes-sigs/scheduler-plugins/tree/master/pkg/noderesourcetopology).
+The goal is to store and update network costs for nodes across regions, zones, or within zones.  
+
+As an initial design, network weights can be manually defined in a single NetworkTopology CR where network costs between
+zones and between regions are specified.
+
+As a separate project, we are providing a [networkTopology controller](https://github.com/jpedro1992/network-topology-controller) in the future. The controller is designed to react to node additions, updates, or removals, and also update the network weights based on latency measurements. 
+
+A [netperf component](https://github.com/jpedro1992/pushing-netperf-metrics-to-prometheus/tree/main) has been developed to execute netperf tests based on the nodes available in the infrastructure. It allows estimating the latency between nodes in the cluster, especially different latency percentiles (i.e., 50th, 90th, and 99th percentile). Latency probing for all pairwise nodes is costly for large-scale clusters. Hence, the initial design assumes that the intra-zone cost between nodes is homogeneous and only the latencies between zones and regions need to be measured.
+
+These measurements are stored and updated in a configmap as key-value pairs with **origin** and **destination** as labels. Then, the networkTopology controller accesses the configmap to extract the netperf measurements and calculates accurate network costs across regions and zones in the cluster. The networkTopology controller can then dynamically update the CR accordingly, so scheduling plugins always use the updated network weights instead of the one-time manually configured weights. 
```

Comment:
The networkTopology controller updates the CR directly but it can extract netperf measurements from a configmap just as an optional initialization. As mentioned, the periodical probing via the netperf component should be limited and preferably avoided in large-scale clusters. 

We believe one-time probing between a single pair of nodes is sufficient if nodes within a zone have similar connections. So, the data in the configmap is not frequently modified (e.g., once a day). 

Essentially, the data is retrieved from the configmap by the controller and then the CR is updated accordingly. The CR will own the manual-defined costs and the calculated ones by the controller with the measurements available in the configmap, so users can decide which ones should be used by the NetworkOverhead plugin. 

Users might prefer to use only manual defined costs and do not need to deploy the netperf component or the networkTopology controller.

Reply:
So you're suggesting the controller may use different net perf measurement tools as the data provider? If that's the case, I think that should be defined as a startup parameter of the controller, instead of a CR field.

Reply:
The controller is generic in its design so that users can decide which metrics they prefer to use. 

We favor netperf measurements between cluster nodes through a netperf component, but other users might prefer to establish costs related to other metrics (e.g., throughput, packet loss, etc). 

In fact, users can define these costs manually without needing to deploy tools like netperf. This is very practical especially for datacenter use cases since only the provider knows the underlying network topology and can manually establish costs between nodes/racks based on their zone/region labels. This is the main reason we favor initializing these via config maps rather than placing arguments to allow different tools.

We included a `configmapName` field in the CR to tell the controller where the metrics will be available for the costs update for that particular CR. Several CRs might be deployed in the cluster, so different config maps can represent different metrics for different CRs. 

The NetworkOverhead plugin has a `networkTopology` argument so that users can select the networkTopology CR to be used by the plugin.

Nevertheless, we can adapt the design and the config map name can be a parameter of the controller.